### DISCUSSION TITLE
Which API permissions are needed?

### DISCUSSION THREAD

**Discussion by ilovelinux** [lowlighter/metrics#1092] [Q&A] [answered]
Hi,

I like this plugin. I was about to use it but I don't want to create a full-privileged API.
I looked into the documentation but I haven´t been able to find anything.

Related to #1050 and partially solved by 814acf63c61642a5e576e7ff14690b906d686cca but I would like a more in-depth response

**Comment by lowlighter** [comment, marked as answer]
Hi 👋 !

For personal access tokens (i.e. tokens used to fetch data), the revelant documentation is [here](https://github.com/lowlighter/metrics/blob/master/.github/readme/partials/documentation/setup/action.md#1%EF%B8%8F-create-a-github-personal-token):
> A a general rule, the following scopes may be required:
>
> - `public_repo` for some plugins
> - `read:org` for all organizations related metrics
> - `repo` for all private repositories related metrics
>   -  `read:user` for some private repositories related metrics
> - `read:packages` for some packages related metrics
> - `gist` for publishing renders to gists instead of a repository

If you click on a plugin documentation in the [plugin list](https://github.com/lowlighter/metrics#-plugins), you'll be able to see a table that tell which scopes are required to make it work, like below:

![image](https://user-images.githubusercontent.com/22963968/173263128-c4017d54-56cf-44c5-ad86-1e829c0e63b3.png)

In the example above, the personal access token should have `read:user` and `read:org` scopes, while the `read:packages`/`repo` are optional and could result in additional stats.

Concerning comitter tokens (i.e. tokens used to publish results), it mostly depends on [`output_action`](https://github.com/lowlighter/metrics/blob/master/source/plugins/core/README.md#-configuring-output-action).

Unfortunately the [GitHub doc](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token) about this is not very extensive and I'm not sure what each scope means as they're named differently from access tokens:
![image](https://user-images.githubusercontent.com/22963968/173263487-2a213136-1704-42ea-93f5-2adb3872e966.png)

However, I'm pretty sure you'll just need `contents: write` if you use any output that will push to said repository (either commit/merge), and `pull-requests: write` if you use the pull request mode. If you use the `gist` feature, then I think no permissions are required, except that the personal access token need to have `gists` scope since it'll write in them

Hope this helps, and thanks for using metrics 😄 !
